### PR TITLE
Support of new BalanceType PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR and PROPORTIONAL_TO_GENERATION_REMAINING_MARGIN

### PR DESCRIPTION
**Pull request by caioluke**
Requires upcoming version of powsybl-core to work
See changes from commit -> `add two new BalanceTypes ...` 9df191f
Other changes come from integration branch

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Adds support to PROPORTIONAL_TO_GENERATION_PARTICIPATION_FACTOR and PROPORTIONAL_TO_GENERATION_REMAINING_MARGIN in balanceType

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:
Relates to
https://github.com/powsybl/powsybl-core/pull/2402
https://github.com/powsybl/powsybl-core/pull/2435

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by caioluke** [comment]
> @caioluke be careful your branch seems to be very far from main! You should rebase you dev on main.

I have rebased it on integration core 5.1.0 snapshot yesterday to be able to use the features I needed from it

File: src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java [hunk 1]
```diff
@@ -27,6 +27,8 @@ public final class LfBatteryImpl extends AbstractLfGenerator {
 
     private double droop;
 
+    private double participationFactor = 0.0;
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java:30]: consistency
```suggestion
    private double droop;

    private double participationFactor;
```

File: src/main/java/com/powsybl/openloadflow/network/LfGenerator.java [hunk 2]
```diff
@@ -67,6 +67,10 @@ default double getDroop() {
         return 0;
     }
 
+    default double getParticipationFactor() {
+        return 0.0;
+    }
+
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/LfGenerator.java:73]: consistency
```suggestion
        return 0;
    }

    default double getParticipationFactor() {
        return 0;
    }

```

File: src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java [hunk 3]
```diff
@@ -35,10 +37,13 @@ private LfBatteryImpl(Battery battery, LfNetwork network, LfNetworkParameters pa
...
         if (activePowerControl != null) {
-            participating = activePowerControl.isParticipate() && activePowerControl.getDroop() != 0;
-            if (activePowerControl.getDroop() != 0) {
+            participating = activePowerControl.isParticipate();
+            if (!Double.isNaN(activePowerControl.getDroop())) {
                 droop = activePowerControl.getDroop();
             }
+            if (!Double.isNaN(activePowerControl.getParticipationFactor())) {
+                participationFactor = activePowerControl.getParticipationFactor();
+            }
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java:46]: we should also verify provided value is positive I think

File: src/main/java/com/powsybl/openloadflow/network/impl/LfGeneratorImpl.java [hunk 4]
```diff
@@ -30,19 +30,23 @@ public final class LfGeneratorImpl extends AbstractLfGenerator {
...
-            boolean withDroop = !Double.isNaN(activePowerControl.getDroop()) && activePowerControl.getDroop() != 0;
-            participating = activePowerControl.isParticipate() && withDroop;
-            if (withDroop) {
+            participating = activePowerControl.isParticipate();
+            if (!Double.isNaN(activePowerControl.getDroop())) {
                 droop = activePowerControl.getDroop();
             }
+            if (!Double.isNaN(activePowerControl.getParticipationFactor())) {
+                participationFactor = activePowerControl.getParticipationFactor();
+            }
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfGeneratorImpl.java:49]: we should also verify provided value is positive I think

File: src/main/java/com/powsybl/openloadflow/network/util/GenerationActivePowerDistributionStep.java [hunk 5]
```diff
@@ -45,7 +47,7 @@ public List<ParticipatingElement> getParticipatingElements(Collection<LfBus> bus
         return buses.stream()
                 .filter(bus -> bus.isParticipating() && !bus.isDisabled() && !bus.isFictitious())
                 .flatMap(bus -> bus.getGenerators().stream())
-                .filter(generator -> generator.isParticipating() && getParticipationFactor(generator) != 0)
+                .filter(generator -> isParticipating(generator) && getParticipationFactor(generator) != 0)
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/util/GenerationActivePowerDistributionStep.java:50]: we should also verify provided value is _strictly_ positive I think
```suggestion
                .filter(generator -> isParticipating(generator) && getParticipationFactor(generator) > 0)
```

**Review by jeandemanged — CHANGES_REQUESTED** [review]

**Comment by jeandemanged** [comment]
sonar report lots of duplicate (relative to the modified code)

but I am not so sure of the best way to get rid of the code duplicates without compromising code readability ... should we really do something about it ?

File: src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java [hunk 6]
```diff
@@ -98,6 +103,11 @@ public double getDroop() {
         return droop;
     }
 
+    @Override
+    public double getParticipationFactor() {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java:107]: I don't understand what will be the value returned if in the extension the participation is negative. We have a default value for the droop. Zero?

Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/network/impl/LfBatteryImpl.java:107]: yes, default value of (primitive) `double` is zero

File: src/main/java/com/powsybl/openloadflow/network/util/GenerationActivePowerDistributionStep.java [hunk 7]
```diff
@@ -105,17 +107,38 @@ public double run(List<ParticipatingElement> participatingElements, int iteratio
...
+        // then depending on participation type, a generator may be found to not participate
+        switch (participationType) {
+            case MAX:
+                return generator.getDroop() != 0;
+            case PARTICIPATION_FACTOR:
+                return generator.getParticipationFactor() > 0;
+            case TARGET:
+            case REMAINING_MARGIN:
+                // nothing more to do here
+                // the check whether TargetP is within Pmin-Pmax range was already included in generator.isParticipating()
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/util/GenerationActivePowerDistributionStep.java:138]: Fix the comment here maybe?

Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/network/util/GenerationActivePowerDistributionStep.java]: was unclear indeed, fixed this way:
```
                // nothing more to do here: the check whether TargetP is within Pmin-Pmax range
                // was already made in AbstractLfGenerator#checkActivePowerControl
                // whose result is reflected in generator.isParticipating()
```

**Review by annetill — APPROVED** [review]

**Review by jeandemanged — APPROVED** [review]

**Pull request merged**